cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java

```diff
@@ -46,12 +48,15 @@ final class ArtifactClassLoaderFactory {
   private static final Logger LOG = LoggerFactory.getLogger(ArtifactClassLoaderFactory.class);
 
   private final CConfiguration cConf;
-  private final ProgramRunnerClassLoaderFactory programRunnerClassLoaderFactory;
+  @Nullable
+  private final ProgramRuntimeProviderLoader programRuntimeProviderLoader;
   private final File tmpDir;
 
-  ArtifactClassLoaderFactory(CConfiguration cConf, ProgramRunnerClassLoaderFactory programRunnerClassLoaderFactory) {
+  @VisibleForTesting
+  ArtifactClassLoaderFactory(CConfiguration cConf,
```

Comment:
It's better to have two constructors. One is this one without the annotation. The other is only taking cConf as parameter and annotate it for testing only.

Reply:
Fixed